File: src/main/java/com/example/demo/controller/AuthController.java
package com.example.demo.controller;

import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;

@Controller
public class AuthController {

    @GetMapping("/login")
    public String login(
            @RequestParam(name = "logout", required = false) Boolean logout,
            @RequestParam(name = "error", required = false) Boolean error,
            @RequestParam(name = "username", required = false) String username,
            @RequestParam(name = "message", required = false) String message,
            Model model
    ) {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        boolean isAuthenticated = auth != null && auth.isAuthenticated() && !(auth instanceof AnonymousAuthenticationToken);
        if (isAuthenticated && (logout == null || !logout) && (error == null || !error)) {
            return "redirect:/tools";
        }

        model.addAttribute("currentScreen", "ツール");
        model.addAttribute("selectedScreenName", "ツール");
        model.addAttribute("currentScreenPath", "toolsList");

        model.addAttribute("showLoginModal", true);
        if (Boolean.TRUE.equals(logout)) {
            model.addAttribute("loginMessage", message != null ? message : "ログアウトしました");
        }
        if (Boolean.TRUE.equals(error)) {
            String errorMessage = message != null ? message : "ユーザー名またはパスワードが正しくありません";
            model.addAttribute("loginError", errorMessage);
            if (username != null && !username.isBlank()) {
                model.addAttribute("loginUsername", username);
            }
        }

        return "main";
    }
}

File: src/main/kotlin/com/example/demo/AuthController.kt
package com.example.demo

import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam

@Controller
class AuthController {
    @GetMapping("/login")
    fun login(
        @RequestParam(name = "logout", required = false) logout: Boolean?,
        @RequestParam(name = "error", required = false) error: Boolean?,
        @RequestParam(name = "username", required = false) username: String?,
        @RequestParam(name = "message", required = false) message: String?,
        model: Model
    ): String {
        // 既にログイン済みで、明示的なログアウト/エラー指定がない場合はツールへ
        val auth: Authentication? = SecurityContextHolder.getContext().authentication
        val isAuthenticated = auth != null && auth.isAuthenticated && auth !is AnonymousAuthenticationToken
        if (isAuthenticated && (logout != true) && (error != true)) {
            return "redirect:/tools"
        }

        // 背景はツール画面相当を表示
        model.addAttribute("currentScreen", "ツール")
        model.addAttribute("selectedScreenName", "ツール")
        model.addAttribute("currentScreenPath", "toolsList")

        // ログインモーダル表示用フラグ/メッセージ
        model.addAttribute("showLoginModal", true)
        if (logout == true) {
            model.addAttribute("loginMessage", message ?: "ログアウトしました")
        }
        if (error == true) {
            val errorMessage = message ?: "ユーザー名またはパスワードが正しくありません"
            model.addAttribute("loginError", errorMessage)
            if (!username.isNullOrBlank()) model.addAttribute("loginUsername", username)
        }

        return "main"
    }
}
